### PR TITLE
[Work-in-Progress] Download a certificate of completion(#76)

### PR DESCRIPTION
**Pull request by Deep-Chill**
Allows downloading a certificate of completion at the /sequence_slug/certificate URL after all elements in the sequence have been viewed.
- `settings.py`: Updated logger to only show warning messages and not the font subsetting and optimization messages that show up due to fontTools

### CODE REVIEW

File: pages/urls/views/elements.py [hunk 1]
```diff
@@ -24,9 +24,12 @@
 
 from ... import settings
 from ...compat import re_path
-from ...views.elements import PageElementView
+from ...views.elements import PageElementView, CertificateDownloadView
 
 urlpatterns = [
+    re_path(r'^(?P<sequence_slug>%s)/certificate' % settings.SLUG_RE,
```
Comment (smirolo) [inline pages/urls/views/elements.py]: Add a '/' to the URL. By convention, API endpoints: no trailing '/', Pages/Downloads: trailing '/'.

File: pages/views/elements.py [hunk 2]
```diff
@@ -218,3 +226,44 @@ def get_context_data(self, **kwargs):
                         kwargs=url_kwargs),
                 })
         return context
+
+
+class CertificateDownloadView(LoginRequiredMixin, DetailView):
```
Comment (smirolo) [inline pages/views/elements.py]: I am not sure we want to have a `LoginRequiredMixin` here. The djaodjin-pages app is used within other projects that handle role-based access control (RBAC) through URL decorators.

File: pages/views/elements.py [hunk 3]
```diff
@@ -218,3 +226,44 @@ def get_context_data(self, **kwargs):
...
+            sequence=sequence,
+            user=self.request.user)
+        enumerated_elements = EnumeratedElements.objects.filter(
+            sequence=sequence)
+        user_enumerated_progress = EnumeratedProgress.objects.filter(
+            progress=sequence_progress)
+
+        has_completed_sequence = True
+        for element in enumerated_elements:
+            user_element_progress = user_enumerated_progress.filter(rank=element.rank).first()
```
Comment (smirolo) [inline pages/views/elements.py]: Relies on N (maximum number of elements) queries to the database for comparing two sequences.

File: pages/views/elements.py [hunk 4]
```diff
@@ -218,3 +226,44 @@ def get_context_data(self, **kwargs):
...
+            if not user_element_progress or user_element_progress.viewing_duration < element.min_viewing_duration:
+                has_completed_sequence = False
+                break
+
+        if has_completed_sequence:
+            last_element = enumerated_elements.last().page_element
```
Comment (smirolo) [inline pages/views/elements.py:274]: Based on the code above, I guess we rely on the last (certificate) element to have `min_viewing_duration == 0`.

File: pages/views/elements.py [hunk 5]
```diff
@@ -218,3 +226,44 @@ def get_context_data(self, **kwargs):
...
+            context.update({
+                'user': self.request.user,
+                'sequence': sequence,
+                'current_date': date.today(),
```
Comment (smirolo) [inline pages/views/elements.py]: Interesting. We rely on the current Web server date.
1. It could be important to pass the timezone as a query parameter just to make sure a one off day does not trigger e-mail and calls to tech support.
2. The date should most likely be fixed so that two successive downloads do not return different date. This might be important for CEUs and other certificate that expire, triggering taking the course again.

**Review by smirolo — COMMENTED** [review]
I will have to write tests first to check various corner cases before merging the PR. Thank you!

File: pages/views/elements.py [hunk 6]
```diff
@@ -246,19 +246,36 @@ def get_context_data(self, **kwargs):
...
-                has_completed_sequence = False
-                break
+        last_element = enumerated_elements.last().page_element
+        is_last_element_certificate = hasattr(last_element, 'certificate')
+
+        if is_last_element_certificate:
+            # Exclude last element from the check, if it is a certificate
+            enumerated_elements = enumerated_elements.exclude(pk=enumerated_elements.last().pk)
+
+        completed_elements_subquery = user_enumerated_progress.filter(
```
Comment (smirolo) [inline pages/views/elements.py:256]: Really cool subquery using `OuterRef`. Does that change the minimum Django version required to run djaodjin-pages?

**Review by smirolo — APPROVED** [review]
Looks really good. Thank you.
